### PR TITLE
Canonicalise install path for comparison

### PR DESCRIPTION
**Pull request by asuffield**
Currently, if the install directory is placed on a path with a symlink
in it, this function will always believe there is a new version,
because it compares realpath() of the symlink with its intended
resolution.

This change realpath()s both the symlink and its intended resolution,
which should give correct behaviour even in the presence of symlinks.

### CODE REVIEW

**Comment by irengrig** [comment]
It does not compile (error from CI): 

Use --sandbox_debug to see verbose messages from the sandbox
--
  | src/main/cpp/blaze.cc: In function 'void blaze::EnsureCorrectRunningVersion(blaze::BlazeServer*)':
  | src/main/cpp/blaze.cc:1243:63: error: cannot convert 'std::__cxx11::string {aka std::__cxx11::basic_string<char>}' to 'const char*' for argument '1' to 'std::__cxx11::string blaze_util::MakeCanonical(const char*)'
  | blaze_util::MakeCanonical(globals->options->install_base);
  | ^

**Comment by asuffield** [comment]
@irengrig Oops, fixed

**Comment by asuffield** [comment]
For maximum fun, this is not right for posix, because windows has different semantics. Fix incoming...

**Comment by asuffield** [comment]
Details: on windows, ReadDirectorySymlink does a RealPath on the results. On posix, it does a single readlink(). The only way to get consistent results is to pass both sides through MakeCanonical before comparing them.

I think it's right now.

File: src/main/cpp/blaze.cc [hunk 1]
```diff
@@ -1236,8 +1236,15 @@ static void EnsureCorrectRunningVersion(BlazeServer *server) {
   string prev_installation;
   bool ok =
       blaze_util::ReadDirectorySymlink(installation_path, &prev_installation);
+  // The windows vesion of ReadDirectorySymlink will RealPath the
```
Comment (laszlocsomor) [inline src/main/cpp/blaze.cc]: This comment may get stale with time. Maybe start with "As of 2019-06-03, ..." or reference this PR.

**Comment by laszlocsomor** [comment]
1. How exactly do I repro the issue? (Tried on Linux with `--output_user_root=/tmp/foo/sym2/bar` where `sym2` -> `sym1` -> `real`, but that didn't work).
2. Could you add a test? Maybe in `//src/test/shell/integration:startup_options_test`, or add a new test (maybe based on startup_options_test).

**Comment by asuffield** [comment]
The bug only happens on windows - ReadDirectorySymlink is readlink() on posix and realpath() on windows. Should be easy to reproduce there with the same arrangement.

I'll try to get a test working.

**Comment by ulfjack** [comment]
Ping @asuffield

**Comment by meisterT** [comment]
What's the status of this PR?

**Comment by meisterT** [comment]
ping @asuffield do you plan to pick this up?

**Comment by philwo** [comment]
Closing due to inactivity, please feel free to reopen if you want to continue working on this.

**Pull request closed without merging**